provider/openstack/provider.go

```diff
@@ -1233,6 +1233,12 @@ func (e *Environ) startInstance(
 			var serverDetail *nova.ServerDetail
 			serverDetail, err = waitForActiveServerDetails(client, server.Id, 5*time.Minute)
 			if err != nil || serverDetail == nil {
+				// If we got an error back (eg. StillBuilding)
+				// we need to terminate the instance before
+				// retrying to avoid leaking resources.
+				if termErr := e.terminateInstances(ctx, []instance.Id{instance.Id(server.Id)}); termErr != nil {
+					logger.Debugf("Failed to delete instance in BUILD state, %q", termErr)
```

Comment:
Suggest including server.ID, to aid in cleanup on failure.

It's not necessarily true that the machine is in the BUILD state here.

Reply:
Good idea. If you get here the machine is either in the BUILD state and we timed out or we got an API error. I will revise the message tomorrow and remove the BUILD part.